File: taste-savant-ios/Taste Savant/Neighborhood.m
//
//  Neighborhood.m
//  Taste Savant
//
//  Created by Joe Gallo on 2/18/13.
//  Copyright (c) 2013 Taste Savant. All rights reserved.
//

#import "Neighborhood.h"

@implementation Neighborhood

- (id)initWithId:(int)id name:(NSString *)name {
    return [self initWithId:id name:name parentName:nil];
}

- (id)initWithId:(int)id name:(NSString *)name parentName:(NSString *)parentName {
    self = [super init];
    if (self) {
        self.id = id;
        self.name = name;
        self.parentName = parentName;
    }
    return self;
}

- (id)initWithDict:(NSDictionary *)dict {
    self = [super init];
    if (self) {
        self.id = [dict[@"id"] intValue];
        self.name = [dict objectForKeyNotNull:@"name"];
        self.parentName = [dict objectForKeyNotNull:@"parent"];
        self.borough = [dict objectForKeyNotNull:@"borough"];
        self.city = [dict objectForKeyNotNull:@"city"];
        self.children = [dict objectForKeyNotNull:@"children"];
    }
    return self;
}

// encode
- (void)encodeWithCoder:(NSCoder *)encoder {
    [encoder encodeInt:self.id forKey:@"id"];
    [encoder encodeObject:self.name forKey:@"name"];
    [encoder encodeInt:self.parentId forKey:@"parentId"];
    [encoder encodeObject:self.parentName forKey:@"parentName"];
    [encoder encodeObject:self.borough forKey:@"borough"];
    [encoder encodeObject:self.city forKey:@"city"];
    [encoder encodeObject:self.children forKey:@"children"];
}

// decode
- (id)initWithCoder:(NSCoder *)decoder {
    self = [super init];
    if (self) {
        self.id = [decoder decodeIntForKey:@"id"];
        self.name = [decoder decodeObjectForKey:@"name"];
        self.parentId = [decoder decodeIntForKey:@"parentId"];
        self.parentName = [decoder decodeObjectForKey:@"parentName"];
        self.borough = [decoder decodeObjectForKey:@"borough"];
        self.city = [decoder decodeObjectForKey:@"city"];
        self.children = [decoder decodeObjectForKey:@"children"];
    }
    return self;
}

+ (Neighborhood *)currentLocation {
    Neighborhood *neighborhood = [[Neighborhood alloc] initWithId:[NSNumber numberWithInt:-1]
                                                             name:@"Current Location"
                                                           parentName:nil];
    return neighborhood;
}

- (id)copyWithZone:(NSZone *)zone {
    Neighborhood *copy = [[[self class] allocWithZone:zone] init];
    if (copy) {
        copy.id = self.id;
        copy.name = self.name;
        copy.parentId = self.parentId;
        copy.parentName = self.parentName;
        copy.borough = self.borough;
        copy.city = self.city;
        copy.children = self.children;
    }
    
    return copy;
}

- (NSComparisonResult)compare:(Neighborhood *)otherObject {
    return [self.name compare:otherObject.name];
}

- (BOOL)isEqualToNeighborhood:(Neighborhood *)aNeighborhood {
    if (self == aNeighborhood)
        return YES;
    if (![(id)[NSNumber numberWithInt:[self id]] isEqual:[NSNumber numberWithInt:[aNeighborhood id]]])
        return NO;
    return YES;
}

- (BOOL)isEqual:(id)other {
    if (other == self)
        return YES;
    if (!other || ![other isKindOfClass:[self class]])
        return NO;
    return [self isEqualToNeighborhood:other];
}

- (NSUInteger)hash {
    NSUInteger prime = 31;
    NSUInteger result = 1;
    result = prime * result + [[NSNumber numberWithInt:self.id] hash];
    return result;
}

- (NSString *)description {
    return [NSString stringWithFormat:
            @"%@ : %@ : %@ (%d)",
            self.city, self.parentName, self.name, self.id];
}

@end

File: nooklyn/nklyn-ios/Nooklyn/Neighborhood.swift
//
//  Neighborhood.swift
//  Nooklyn
//
//  Created by Joe Gallo on 6/2/15.
//  Copyright (c) 2016 Nooklyn. All rights reserved.
//

class Neighborhood: NSObject, NSCoding, NSCopying {
    var id = ""
    var name = ""
    var imageURL = ""
    var featured = false
    var latitude = 0.0
    var longitude = 0.0
    var activeListingCount = 0
    var locationCategoryCount = 0
    var regionId = ""
    var region: Region!
    var locations = [Location]()
    
    override init() {
        super.init()
    }
    
    init(dict: NSDictionary) {
        if let id = dict["id"] as? String { self.id = id }
        
        if let attributes = dict["attributes"] as? NSDictionary {
            if let name = attributes["name"] as? String { self.name = name }
            if let imageURL = attributes["image"] as? String { self.imageURL = imageURL }
            if let featured = attributes["featured"] as? Bool { self.featured = featured }
            if let latitude = attributes["latitude"] as? Double { self.latitude = latitude }
            if let longitude = attributes["longitude"] as? Double { self.longitude = longitude }
            if let activeListingCount = attributes["active-listing-count"] as? Int { self.activeListingCount = activeListingCount }
            if let locationCategoryCount = attributes["location-category-count"] as? Int { self.locationCategoryCount = locationCategoryCount }
        }
        
        if let relationships = dict["relationships"] as? NSDictionary {
            // region
            if let regionDict = relationships["region"] as? NSDictionary {
                if let regionData = regionDict["data"] as? NSDictionary {
                    if let regionId = regionData["id"] as? String {
                        self.regionId = regionId
                    }
                }
            }
        }
    }
    
    // MARK: - NSCoding
    
    required convenience init?(coder decoder: NSCoder) {
        self.init()
        
        if let id = decoder.decodeObjectForKey("id") as? String { self.id = id }
        if let name = decoder.decodeObjectForKey("name") as? String { self.name = name }
        if let imageURL = decoder.decodeObjectForKey("imageURL") as? String { self.imageURL = imageURL }
        self.featured = decoder.decodeBoolForKey("featured")
        self.latitude = decoder.decodeDoubleForKey("latitude")
        self.longitude = decoder.decodeDoubleForKey("longitude")
        self.activeListingCount = decoder.decodeIntegerForKey("activeListingCount")
        self.locationCategoryCount = decoder.decodeIntegerForKey("locationCategoryCount")
        if let regionId = decoder.decodeObjectForKey("regionId") as? String { self.regionId = regionId }
        if let region = decoder.decodeObjectForKey("region") as? Region { self.region = region }
        if let locations = decoder.decodeObjectForKey("locations") as? [Location] { self.locations = locations }
    }
    
    func encodeWithCoder(coder: NSCoder) {
        coder.encodeObject(self.id, forKey: "id")
        coder.encodeObject(self.name, forKey: "name")
        coder.encodeObject(self.imageURL, forKey: "imageURL")
        coder.encodeBool(self.featured, forKey: "featured")
        coder.encodeDouble(self.latitude, forKey: "latitude")
        coder.encodeDouble(self.longitude, forKey: "longitude")
        coder.encodeInt(Int32(self.activeListingCount), forKey: "activeListingCount")
        coder.encodeInt(Int32(self.locationCategoryCount), forKey: "locationCategoryCount")
        coder.encodeObject(self.regionId, forKey: "regionId")
        coder.encodeObject(self.region, forKey: "region")
        coder.encodeObject(self.locations, forKey: "locations")
    }
    
    // MARK: - NSCopying
    
    func copyWithZone(zone: NSZone) -> AnyObject {
        let copy = Neighborhood()
        copy.id = id
        copy.name = name
        copy.imageURL = imageURL
        copy.featured = featured
        copy.latitude = latitude
        copy.longitude = longitude
        copy.activeListingCount = activeListingCount
        copy.locationCategoryCount = locationCategoryCount
        copy.regionId = regionId
        copy.region = region
        copy.locations = locations
        return copy
    }
    
    // MARK: - Compare
    
    override func isEqual(other: AnyObject?) -> Bool {
        if let other = other as? Neighborhood {
            return id == other.id
        } else {
            return false
        }
    }
    
    override var hash: Int {
        return id.hashValue
    }
}

func ==(lhs: Neighborhood, rhs: Neighborhood) -> Bool {
    return lhs.id == rhs.id
}
